### PR TITLE
[11.0] product_internal_reference_unique

### PR DESCRIPTION
**Pull request by agyamuta**
Correspondence in reference to issue https://github.com/OCA/product-attribute/issues/331

### CODE REVIEW

**Comment by max3903** [comment]
@agyamuta Can you provide a demo data file with an internal reference for the demo products that doesn't have one? Thanks

**Comment by pedrobaeza** [comment]
We are calling everywhere "internal reference" as "code", so I think you can name your module as *product_code_unique*

**Review by max3903 — APPROVED** [review]

**Comment by max3903** [comment]
@agyamuta Can you apply the same logic as in #334 with the pre_init_hook?

**Comment by agyamuta** [comment]
@max3903 Applied logic as in #334 . Thanks.

File: product_code_unique/__init__.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# Copyright (C) 2018 - TODAY, Open Source Integrators
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from . import models
+
+
+def pre_init_product_code(cr):
```
Comment (feketemihai) [inline product_code_unique/__init__.py:7]: You will have to treat templates as weel as the other module..

**Comment by feketemihai** [comment]
Some tests would help..

**Review by feketemihai — CHANGES_REQUESTED** [review]

File: product_code_unique/README.rst [hunk 2]
```diff
@@ -0,0 +1,60 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+   :alt: License: AGPL-3
+
+==================================
+Unique Product Internal Reference
+==================================
```
Comment (JayVora-SerpentCS) [inline product_code_unique/README.rst]: Need to remove one '=' to align with the last 'e' of reference

File: product_code_unique/models/product.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models
+
+
+class ProductProduct(models.Model):
+    _inherit = 'product.product'
+
+    default_code = fields.Char('Internal Reference', index=True)
+
```
Comment (JayVora-SerpentCS) [inline product_code_unique/models/product.py]: Was it needed to write the field definition again ?Cannot we add only the constraint?

Comment (agyamuta) [reply product_code_unique/models/product.py]: Thanks for pointing out. Tested and it works the same. This is corresponded.  Unlike with lower odoo version if not redefined changes wont take effect.

File: product_code_unique/__init__.py (hunk 1 above)
Comment (agyamuta) [reply product_code_unique/__init__.py:7]: Product Templates are not mandatory but required when it has by default 1 variant only. So no need to initialize the value in this case.

Comment (agyamuta) [reply product_code_unique/__init__.py:7]: image

Comment (max3903) [reply product_code_unique/__init__.py:7]: @feketemihai 

FYI, @agyamuta will change the method to find and remove duplicate default_code in product_product.

We don't need to touch product_template as the default_code of a product template with many variants will be empty:
https://github.com/odoo/odoo/blob/11.0/addons/product/models/product_template.py#L256

The uniqueness constraint is maintained with empty values.

File: product_code_unique/models/product.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
+# Copyright (C) 2018 - TODAY, Open Source Integrators
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models
```
Comment (JayVora-SerpentCS) [inline product_code_unique/models/product.py]: remove fields, as its unused

File: product_code_unique/README.rst [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+This module adds a constraint on the internal reference of the product 
+to make it unique across the database.
+
+Usage
+=====
+* Unable to save a product when a new internal reference or default_code value is
+  the same with an existing record.
+* A pre_init_hook process is initiated when there exist records without an internal reference(default_code).
+  A default value is generated to populate empty field as a temporary value.
+
```
Comment (JayVora-SerpentCS) [inline product_code_unique/README.rst:18]: Add Runbot link.

**Review by JayVora-SerpentCS — CHANGES_REQUESTED** [review]

**Comment by max3903** [comment]
@JayVora-SerpentCS @feketemihai Can you update your review please?

**Review by JayVora-SerpentCS — APPROVED** [review]

**Pull request merged**